gio-subclass/tests/simple_action.rs

```diff
@@ -0,0 +1,391 @@
+            self.parameter_type.borrow().clone()
+        }
+
+        fn get_state(&self, action: &gio::Action) -> Option<glib::Variant> {
+            self.state.borrow().clone()
+        }
+
+        fn get_state_hint(&self, action: &gio::Action) -> Option<glib::Variant> {
+            self.state_hint.borrow().clone()
+        }
+
+        fn get_state_type(&self, action: &gio::Action) -> Option<glib::VariantType> {
+            self.state_type.borrow().clone()
+        }
+    }
+
+    struct SimpleActionStatic;
+
+    impl ImplTypeStatic<Object> for SimpleActionStatic {
+        fn get_name(&self) -> &str {
+            "SimpleAction"
+        }
+
+        fn new(&self, obj: &Object) -> Box<ObjectImpl<Object>> {
+            SimpleAction::init(obj)
+        }
+
+        fn class_init(&self, klass: &mut ObjectClass) {
+            SimpleAction::class_init(klass);
+        }
+
+        fn type_init(&self, token: &TypeInitToken, type_: glib::Type) {
+            register_action(token, type_, self);
+        }
+    }
+
+    impl ActionImplStatic<Object> for SimpleActionStatic {
+        fn get_impl<'a>(&self, imp: &'a Box<ObjectImpl<Object>>) -> &'a ActionImpl {
+            imp.downcast_ref::<SimpleAction>().unwrap()
+        }
+    }
+}
+
+glib_wrapper! {
+    pub struct SimpleAction(Object<imp::SimpleAction>):
+        [Object => InstanceStruct<Object>,
+         gio::Action => gio_ffi::GAction];
+
+    match fn {
+        get_type => || imp::SimpleAction::get_type().to_glib(),
+    }
+}
+
+impl SimpleAction {
+    pub fn new<'a, P: Into<Option<&'a glib::VariantTy>>>(
+        name: &str,
+        parameter_type: P,
+    ) -> SimpleAction {
+        use glib::object::Downcast;
+
+        let ty: Option<glib::VariantType> = parameter_type.into().map(ToOwned::to_owned);
+
+        unsafe {
+            glib::Object::new(
+                Self::static_type(),
+                &[("name", &name), ("parameter-type", &ty.to_value())],
+            ).unwrap()
+                .downcast_unchecked()
+        }
+    }
+
+    pub fn new_stateful<'a, P: Into<Option<&'a glib::VariantTy>>>(
+        name: &str,
+        parameter_type: P,
+        state: &glib::Variant,
+    ) -> SimpleAction {
+        use glib::object::Downcast;
+
+        unsafe {
+            glib::Object::new(
+                Self::static_type(),
+                &[
+                    ("name", &name),
+                    ("parameter-type", &parameter_type.into()),
+                    ("state", &state),
+                ],
+            ).unwrap()
+                .downcast_unchecked()
+        }
+    }
+
+    pub fn connect_activate<F: Fn(&Self, Option<glib::Variant>) + 'static>(
+        &self,
+        f: F,
+    ) -> glib::SignalHandlerId {
+        // FIXME: This needs some simplification...
+        //
+        // Get us Send/Sync constraints
+        let f: Box<Fn(&Self, Option<glib::Variant>) + Send + Sync + 'static> = unsafe {
```

Comment:
Fancy! `GVariant` is be definition `Send` and `Sync` (it's immutable after all). That should be fixed in GLib

Reply:
It is needed like it is there, but not because of ```GVariant```. Rather because of ```self``` which is not ```Send``` + ```Sync```. Does it need to be?

Reply:
Can you nonetheless submit a PR for `GVariant`? It will probably appear again sooner or later for real :)

For `Self`, it's only needed because we can't statically know if the signal involves other threads. In this case it is safe what you're doing here.